main/forum/editthread.php

```diff
@@ -0,0 +1,220 @@
+api_protect_course_script(true);
+
+$nameTools = get_lang('ToolForum');
+
+// Unset the formElements in session before the includes function works
+unset($_SESSION['formelements']);
+
+/* Including necessary files */
+
+require_once 'forumconfig.inc.php';
+require_once 'forumfunction.inc.php';
+
+// Are we in a lp ?
+$origin = '';
+if (isset($_GET['origin'])) {
+    $origin = Security::remove_XSS($_GET['origin']);
+}
+
+/* MAIN DISPLAY SECTION */
+$current_forum = get_forum_information($_GET['forum']);
+$current_forum_category = get_forumcategory_information($current_forum['forum_category']);
+
+/* Breadcrumbs */
+
+if (isset($_SESSION['gradebook'])){
+    $gradebook = Security::remove_XSS($_SESSION['gradebook']);
+}
+
+if (!empty($gradebook) && $gradebook == 'view') {
+    $interbreadcrumb[] = array (
+        'url' => '../gradebook/'.Security::remove_XSS($_SESSION['gradebook_dest']),
+        'name' => get_lang('ToolGradebook')
+    );
+}
+
+if (!empty($_GET['gidReq'])) {
+    $toolgroup = intval($_GET['gidReq']);
+    Session::write('toolgroup',$toolgroup);
+}
+
+$threadId = isset($_GET['thread']) ? $_GET['thread'] : false;
+$courseInfo = isset($_GET['cidReq']) ? api_get_course_info($_GET['cidReq']) : false;
+$cId = isset($courseInfo['real_id']) ? $courseInfo['real_id'] : false; 
+
+/* Is the user allowed here? */
+
+// The user is not allowed here if:
+
+// 1. the forumcategory or forum is invisible (visibility==0) and the user is not a course manager
+if (!api_is_allowed_to_edit(false, true) &&
+    (($current_forum_category['visibility'] && $current_forum_category['visibility'] == 0) || $current_forum['visibility'] == 0)
+) {
+    api_not_allowed();
+}
+
+// 2. the forumcategory or forum is locked (locked <>0) and the user is not a course manager
+if (!api_is_allowed_to_edit(false, true) &&
+    (($current_forum_category['visibility'] && $current_forum_category['locked'] <> 0) OR $current_forum['locked'] <> 0)
+) {
+    api_not_allowed();
+}
+
+// 3. new threads are not allowed and the user is not a course manager
+if (!api_is_allowed_to_edit(false, true) &&
+    $current_forum['allow_new_threads'] <> 1
+) {
+    api_not_allowed();
+}
+// 4. anonymous posts are not allowed and the user is not logged in
+if (!$_user['user_id'] AND $current_forum['allow_anonymous'] <> 1) {
+    api_not_allowed();
+}
+
+// 5. Check user access
+if ($current_forum['forum_of_group'] != 0) {
+    $show_forum = GroupManager::user_has_access(
+        api_get_user_id(),
+        $current_forum['forum_of_group'],
+        GroupManager::GROUP_TOOL_FORUM
+    );
+    if (!$show_forum) {
+        api_not_allowed();
+    }
+}
+
+// 6. Invited users can't create new threads
+if (api_is_invitee()) {
+    api_not_allowed(true);
+}
+
+$groupId = api_get_group_id();
+if (!empty($groupId)) {
+    $group_properties = GroupManager :: get_group_properties($groupId);
+    $interbreadcrumb[] = array('url' => '../group/group.php?'.api_get_cidreq(), 'name' => get_lang('Groups'));
```

Comment:
api_get_cidreq() is a heavy/slow function. For repeating occurences, you'd better use $cidreq = api_get_cidreq() and then use $cidreq in the string. This has to be local though, to avoid issues later on. So just for the if/else in this case, for example. After that, if you have another api_get_cidreq() further down (after execution of code), then leave it like that.